iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/procedure/env/RegionMaintainHandler.java

```diff
@@ -661,50 +660,48 @@ public void removeDataNodePersistence(TDataNodeLocation dataNodeLocation) {
     Optional<TDataNodeLocation> newLeaderNode =
         filterDataNodeWithOtherRegionReplica(regionId, originalDataNode);
-
-    if (TConsensusGroupType.DataRegion.equals(regionId.getType())
-        && IOT_CONSENSUS.equals(CONF.getDataRegionConsensusProtocolClass())) {
-      if (newLeaderNode.isPresent()) {
-        configManager
-            .getLoadManager()
-            .forceUpdateConsensusGroupCache(
-                Collections.singletonMap(
-                    regionId,
-                    new ConsensusGroupHeartbeatSample(
-                        System.nanoTime(), newLeaderNode.get().getDataNodeId())));
-        LOGGER.info(
-            "{}, Change region leader finished for IOT_CONSENSUS, regionId: {}, newLeaderNode: {}",
-            REGION_MIGRATE_PROCESS,
-            regionId,
-            newLeaderNode);
+    newLeaderNode.orElseThrow(() -> new ProcedureException("Cannot find the new leader"));
+
+    // ratis needs DataNode to do election by itself
+    if (TConsensusGroupType.SchemaRegion.equals(regionId.getType())
+        || TConsensusGroupType.DataRegion.equals(regionId.getType())
+            && RATIS_CONSENSUS.equals(CONF.getDataRegionConsensusProtocolClass())) {
+      final int MAX_RETRY_TIME = 10;
+      int retryTime = 0;
+      while (true) {
+        TRegionLeaderChangeResp resp =
+            SyncDataNodeClientPool.getInstance()
+                .changeRegionLeader(
+                    regionId, originalDataNode.getInternalEndPoint(), newLeaderNode.get());
+        if (resp.getStatus().getCode() == TSStatusCode.SUCCESS_STATUS.getStatusCode()) {
+          break;
+        }
+        if (retryTime++ > MAX_RETRY_TIME) {
+          throw new ProcedureException("Transfer leader fail");
+        }
+        LOGGER.warn("Call changeRegionLeader fail for the {} time", retryTime);
       }
-
-      return;
     }
 
-    if (newLeaderNode.isPresent()) {
-      TRegionLeaderChangeResp resp =
-          SyncDataNodeClientPool.getInstance()
-              .changeRegionLeader(
-                  regionId, originalDataNode.getInternalEndPoint(), newLeaderNode.get());
-      if (resp.getStatus().getCode() == TSStatusCode.SUCCESS_STATUS.getStatusCode()) {
-        configManager
-            .getLoadManager()
-            .forceUpdateConsensusGroupCache(
-                Collections.singletonMap(
-                    regionId,
-                    new ConsensusGroupHeartbeatSample(
-                        resp.getConsensusLogicalTimestamp(), newLeaderNode.get().getDataNodeId())));
-      }
-      LOGGER.info(
-          "{}, Change region leader finished for RATIS_CONSENSUS, regionId: {}, newLeaderNode: {}",
-          REGION_MIGRATE_PROCESS,
-          regionId,
-          newLeaderNode);
-    }
+    configManager
+        .getLoadManager()
+        .forceUpdateConsensusGroupCache(
+            Collections.singletonMap(
+                regionId,
+                new ConsensusGroupHeartbeatSample(
```

Comment:
should be the TRegionLeaderChangeResp. getConsensusLogicalTimestamp() for ratisconsensus?

Reply:
sure